### PR TITLE
ENH: progress reporting does not stall with PoolMultiThreader

### PR DESCRIPTION
**Pull request by dzenanz**
Implementing the suggestion made in [a forum post](https://discourse.itk.org/t/no-progress-reporting-in-itk5/2630/33?u=dzenanz). Related to #1673.
## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

### CODE REVIEW

File: Modules/Core/Common/src/itkPoolMultiThreader.cxx [hunk 1]
```diff
@@ -273,7 +281,12 @@ PoolMultiThreader ::ParallelizeImageRegion(unsigned int         dimension,
...
       {
         exceptionHandler.TryAndCatch([this, i, &reporter] {
-          m_ThreadInfoArray[i].Future.get();
+          std::future_status status;
+          do
+          {
+            status = m_ThreadInfoArray[i].Future.wait_for(threadCompletionPollingInterval);
+            reporter.Completed(0);
+          } while (status != std::future_status::ready);
           reporter.CompletedPixel();
```
Comment (blowekamp) [inline Modules/Core/Common/src/itkPoolMultiThreader.cxx:290]: Using the TotalProgressReporter, the progress is going to be incremented by the chuck amount. The filter may  also be doing the progress, the two will be added together causing some inaccuracy/difficulties.

Comment (dzenanz) [reply Modules/Core/Common/src/itkPoolMultiThreader.cxx:305]: What is "chuck amount"? I followed your other suggestions.

Comment (blowekamp) [reply Modules/Core/Common/src/itkPoolMultiThreader.cxx:305]: Looks like you reverted to just the progress reported 👍

File: Modules/Core/Common/src/itkPoolMultiThreader.cxx [hunk 2]
```diff
@@ -193,8 +200,16 @@ PoolMultiThreader ::ParallelizeArray(SizeValueType             firstIndex,
...
-      exceptionHandler.TryAndCatch([this, i, &reporter] {
-        m_ThreadInfoArray[i].Future.get();
+      exceptionHandler.TryAndCatch([this, i, &reporter, &filter] {
+        std::future_status status;
+        do
+        {
+          status = m_ThreadInfoArray[i].Future.wait_for(threadCompletionPollingInterval);
+          if (filter)
+          {
+            filter->IncrementProgress(0);
```
Comment (blowekamp) [inline Modules/Core/Common/src/itkPoolMultiThreader.cxx:210]: This fine tweak to consider, but this method always trigger a ProgressReport event. There are some cases where regular updates may be nice ( as this should provide ) or situations where the performance of not emitting extra events would be desired.

**Comment by dzenanz** [comment]
I went with your code suggestions Brad. Platform multi-threader does not deal with progress at all, as it is only meant for backwards compatibility.

Writing a good test for this is not trivial. Perhaps you or @spinicist could do that? If not, let's not let perfect be the enemy of the good 😄

**Comment by spinicist** [comment]
Hello,

Thanks for getting round to this and including me. I'm not working on the project that needs this right at the moment - hopefully will get back to it in a couple of weeks.

I'm wondering what a test for this would look like? Print the progress messages to a string buffer and then check they follow a known pattern?

**Comment by dzenanz** [comment]
Come up with some dummy work (could even just be sleeping), add progress observer, and check that progress is reported a few times for 50%-95% range. A number of threads could be programatically fixed to say 4, and number of work units to say 20. Also explicitly request Pool multi-threader.

**Comment by blowekamp** [comment]
I manually tested this patch with the itkMedianImageFilterTest on a large image with a the Pool and Platform threaders with a variety of different number of threads. The progressed reporting seemed to work very well!

Also SimpleITK's tests all passed when build against this PR.

Thanks for coming up with this approach to improve the regularity or reporting.

**Comment by blowekamp** [comment]
I manually tested this patch with the itkMedianImageFilterTest on a large image with a the Pool and Platform threaders with a variety of different number of threads. The progressed reporting seemed to work very well!

Also SimpleITK's tests all passed when build against this PR.

**Comment by dzenanz** [comment]
Tobias, do you want to add a unit test in this PR? Or perhaps plan to do it in a follow-up PR?

**Comment by spinicist** [comment]
Hi, sadly I have not had time to get back to the project that needs this, and I don't know when I may do so. I very much appreciate your efforts on this and don't want to hold you up so please go ahead and merge, and when I finally get back to ITK I will look at adding a test.

**Pull request merged**